nutt_eric_ndvi.py

```diff
@@ -0,0 +1,662 @@
+
+# In[5]:
+
+
+# In this cell place all of the functions needed to run your notebook
+# You will be graded here on function application, docstrings, efficiency so ensure
+# All functions are placed here!
+
+# Define function to open and clean a single landsat band
+def open_clean_bands(band_path,
+                     valid_range=None):
+    """Open and mask a single landsat band using a pixel_qa layer.
+
+    Parameters
+    -----------
+    band_path : string
+        A path to the array to be opened
+    valid_range : tuple (optional)
+        A tuple of min and max range of values for the data. Default = None
+
+
+    Returns
+    -----------
+    arr : xarray DataArray
+        An xarray DataArray with values that should be masked set to 1 for True (Boolean)
+    """
+    # Open a single band using rioxarray & mask to valid range then apply cloud mask
+    band = (rxr.open_rasterio(band_path, masked=True)
+            .rio.clip(crop_bounds.geometry, from_disk=True)
+            .squeeze())
+
+    # Specify valid range of values
+    if valid_range:
+        mask = ((band <= 0) | (band > 10000))
+        band = band.where(~mask, np.nan)
+
+    return band
+
+
+# Define function to open and mask single landsat band
+def mask_crop_ndvi(all_bands,
+                   crop_bound,
+                   pixel_qa_path,
+                   vals):
+    """Open and mask a single landsat band using a pixel_qa layer.
+
+    Parameters
+    -----------
+    all_bands : list
+        a list containing the xarray objects for landsat  bands 4 and  5
+    crop_bound: geopandas GeoDataFrame
+        A geopandas dataframe to be used to crop the raster data using rasterio mask().
+    pixel_qa: xarray DataArray
+        An xarray DataArray with pixel qa values that have not yet been turned into a mask (0s and 1s)
+    vals: list
+        A list of values needed to create the cloud mask
+
+
+    Returns
+    -----------
+    ndvi_mean : Xarray object
+        an xarray object containing NDVI mean values
+    """
+
+    crop_extent = crop_bounds
+
+    # Open all bands using a loop
+    bands = []
+    for aband in all_bands_path:
+        cleaned_band = open_clean_bands(band_path=aband,
+                                        valid_range=(0, 10000))
+        bands.append(cleaned_band)
+
+    # Clip the cloud mask layer
+    cl_mask = (rxr.open_rasterio(pixel_qa_path, masked=True).squeeze()
+               .rio.clip(crop_bounds.geometry, from_disk=True)
+               .squeeze())
+
+    # Calculate NDVI
+    ndvi_xr = (bands[1]-bands[0]) / (bands[1]+bands[0])
+
+    # Apply cloud mask to NDVI
+    all_masked_values = [328, 392, 840, 904, 1350, 352, 368, 416,
+                         432, 480, 864, 880, 928, 944, 992, 480, 992]
+
+    ndvi_crop = ndvi_xr.where(~cl_mask.isin(all_masked_values))
+
+    # NDVI mean
+    mean_ndvi = ndvi_crop.mean(skipna=True).item()
+
+    return mean_ndvi
```

Comment:
Yes I wanted to explore using nested loops within functions and I think it works okay. Does seem like it may limit future workflow customization, though.